### PR TITLE
Use experimental shopify plugin

### PR DESCRIPTION
**Pull request by sslotsky**
Note that this will currently require an .env file with these keys in order to test:

```
SHOPIFY_STORE_URL=gatsby-swag.myshopify.com
SHOPIFY_ACCESS_TOKEN=
SHOPIFY_ADMIN_API_KEY=
SHOPIFY_ADMIN_PASSWORD=
```

I'll publish another release that accepts these values as plugin options.

## Benchmarking

Tracking build time comparisons here as I go. Might delete later IDK.

### Old plugin

```
success open and validate gatsby-configs - 0.056s
success load plugins - 0.454s
success onPreInit - 0.024s
success initialize cache - 0.005s
success copy gatsby files - 0.019s
success onPreBootstrap - 0.011s
success createSchemaCustomization - 0.004s

gatsby-source-shopify/gatsby-swag starting to fetch data from Shopify

gatsby-source-shopify/gatsby-swag fetched and processed Collection nodes: 489.448ms

gatsby-source-shopify/gatsby-swag fetched and processed ShopPolicy nodes: 479.597ms

gatsby-source-shopify/gatsby-swag fetched and processed Blog nodes: 488.670ms

gatsby-source-shopify/gatsby-swag fetched and processed Article nodes: 587.472ms

gatsby-source-shopify/gatsby-swag fetched and processed Page nodes: 600.137ms
success Downloading remote files - 3.478s - 61/61 17.54/s

gatsby-source-shopify/gatsby-swag fetched and processed Product nodes: 6019.998ms

gatsby-source-shopify/gatsby-swag finished fetching data from Shopify: 6017.212ms
success Checking for changed pages - 0.001s
success source and transform nodes - 6.065s
success building schema - 0.311s
info Total nodes: 366, SitePage nodes: 34 (use --verbose for breakdown)
success createPages - 0.035s
success Checking for changed pages - 0.001s
success createPagesStatefully - 0.037s
success update schema - 0.019s
success write out redirect data - 0.001s
success Build manifest and related icons - 0.083s
success onPostBootstrap - 0.086s
info bootstrap finished - 8.516s
success onPreExtractQueries - 0.001s
success extract queries from components - 0.236s
success write out requires - 0.003s
success run static queries - 3.598s - 3/3 0.83/s
success run page queries - 0.012s - 3/3 240.81/s
success Generating image thumbnails - 176.699s - 1156/1156 6.54/s
```

### New plugin

```
success open and validate gatsby-configs - 0.054s
success load plugins - 0.435s
success onPreInit - 0.022s
success initialize cache - 0.006s
success copy gatsby files - 0.018s
success onPreBootstrap - 0.009s
success createSchemaCustomization - 0.005s
Checked for operations in progress: 569.551ms
Initiated bulk operation query: 376.147ms
Completed bulk operation: 5619.528ms
Created nodes from bulk operation: 2378.194ms
createProductsOperation: 9807.659ms
success Checking for changed pages - 0.001s
success source and transform nodes - 9.839s
success building schema - 0.599s
info Total nodes: 494, SitePage nodes: 36 (use --verbose for breakdown)
success createPages - 0.039s
success Checking for changed pages - 0.000s
success createPagesStatefully - 0.043s
success update schema - 0.017s
success write out redirect data - 0.001s
success Build manifest and related icons - 0.086s
success onPostBootstrap - 0.089s
info bootstrap finished - 12.490s
success onPreExtractQueries - 0.001s
success extract queries from components - 0.301s
success write out requires - 0.007s
success run static queries - 3.510s - 3/3 0.85/s
success run page queries - 0.026s - 3/3 113.65/s
success Generating image thumbnails - 194.428s - 1196/1196 6.15/s
```

### CODE REVIEW

**Comment by sslotsky** [comment]
> Note that this will require an .env file with these keys:

We should probably publish a new version of the plugin that accepts those values as plugin options, but either way, we'll need all the values

**Review by smthomas — DISMISSED** [review]
This looks good to me!

**Comment by sslotsky** [comment]
Check out the benchmarking results. What's odd to me here is that we're not fetching as many data types, but we wind up with more nodes.

It also takes a little bit longer to source nodes in the tests I've run so far. I think this is partly from waiting for the bulk operation to complete. I wouldn't be surprised if this sort of thing slows down builds for smaller Shopify sites by a little bit, while hopefully increasing both speed and reliability for larger sites by a lot.

**Comment by KyleAMathews** [comment]
>It also takes a little bit longer to source nodes in the tests I've run so far. I think this is partly from waiting for the bulk operation to complete.

What are the numbers?

How often do we run the bulk operation? Presumably if we already have one cached locally we just do a delta fetch?

**Comment by sslotsky** [comment]
> > It also takes a little bit longer to source nodes in the tests I've run so far. I think this is partly from waiting for the bulk operation to complete.
> 
> What are the numbers?
> 
> How often do we run the bulk operation? Presumably if we already have one cached locally we just do a delta fetch?

Check out the PR description for lots of numbers, including these

```
success source and transform nodes - 6.065s
```
```
success source and transform nodes - 9.839s
```

And yes, if we have one cached, we'll do the delta fetch. That should be an improvement for sure.

**Comment by sslotsky** [comment]
I added some breakdowns for the timing of `sourceAllNodes` that can also be seen in the results that I pasted in the description:

```
Checked for operations in progress: 569.551ms
Initiated bulk operation query: 376.147ms
Completed bulk operation: 5619.528ms
Created nodes from bulk operation: 2378.194ms
createProductsOperation: 9807.659ms
```

`createProductsOperation` encapsulates the previous four lines. Polling the bulk operation talks ~5.6 seconds. Once that completes it takes ~2.4 seconds to create the nodes.

File: gatsby-config.js [hunk 1]
```diff
@@ -18,13 +18,7 @@ module.exports = {
...
     'gatsby-plugin-sharp',
     'gatsby-transformer-sharp',
-    {
-      resolve: 'gatsby-source-shopify',
-      options: {
-        shopName: 'gatsby-swag',
-        accessToken: process.env.SHOPIFY_ACCESS_TOKEN
-      }
-    },
+    'gatsby-source-shopify-experimental',
```
Comment (sslotsky) [inline gatsby-config.js:21]: Version 1.4.0 accepts the keys as plugin options, I'll update this Monday

**Comment by sslotsky** [comment]
So we have a dilemma here regarding the secrets. The new plugin uses the admin API, and the keys to that API do include some write access for customer data. So we probably do not want to add them to the repo. 

I'm not sure if there's a way to create a new private app with strictly read-only access for the admin API. Might have to enable private app development in order to create a new one?

**Comment by sslotsky** [comment]
Now that the build is succeeding, does anyone want to take a quick look at the preview and make sure nothing looks wrong? Like missing products or anything like that? It looks good to me so if nothing stands out to anyone else we'll merge!

https://build-9c383751-df6e-4c32-aae2-c78588b60df4.gtsb.io/

**Comment by sslotsky** [comment]
After looking over the branch build there are a couple issues that need investigating. 

1. A couple additional products show up
2. Some products show up in a different order
3. Some products feature different images
4. Products with more than one image only seem to process the first image correctly

**Review by smthomas — APPROVED** [review]
😍

**Comment by fk** [comment]
<3

**Pull request merged**